20250828-bucket-eventing-phase-1.md

```diff
@@ -0,0 +1,689 @@
+  - `ObjectRemoved:DeleteMarkerCreated` if the change record’s `mod_type` is `INSERT`, and the `status` new value is either `DeleteMarkerVersioned(5)` or `DeleteMarkerUnversioned(6)`.
+  - `ObjectRemoved:Delete` if the change record’s `mod_type` is `DELETE`, and the `status` old value is either `CommittedUnversioned(3)`, `CommittedVersioned(4)`, `DeleteMarkerVersioned(5)`, or `DeleteMarkerUnversioned(6)`.
+- `userIdentity`: will be skipped as the `objects` table does not keep information who initiated the database transaction
+- `requestParameters`: will be skipped as the `objects` table does not keep information for the source IP of the client who initiated the database transaction
+- `requestParameters`: will be skipped as the `objects` table does not keep information for the client request ID and the host that processed the request
+- `s3SchemaVersion`: will be hardcoded to `1.0`
+- `configurationId`: will be hardcoded to `ObjectEvents`
+  - In a future phase, it will be set to the `ID` from the `PutBucketNotificationConfiguration` request that enabled the bucket eventing
+- `bucket->name`: will be set to the `bucket_name` key of the change record
+- `bucket->ownerIdentity->principalId`: will be set to the public project ID of the bucket. The bucket eventing chore will resolve the public project ID from the `project_id` key (the private project ID) of the change record.
+- `bucket->arn`: will be set to `arn:storj:s3:::<bucket_name>`
+- `object->key`: will be set to the `object_key` key of the change record. There will be no attempt to decrypt the object key. If the object key is encrypted it will be the user's responsibility to decrypt it. We will provide the tooling and documentation.
+- `object->size`: will be set to the new `total_plain_size` value of the change record
+- `object->eTag`: will be skipped for Phase 1
+- `object->versionId`: will be set to the `version` key of the change record and skipped if the bucket is not versioning-enabled.
+- `object->sequencer`:  will be set to the `commit_timestamp` UNIX nanoseconds from the change record, converted to a 16-character, zero-padded, uppercase hex string.
+
+## Disclaimers
+
+### Anti-goals
+
+### Alternatives considered
+
+- We considered the API pods to be the source of bucket events. We concluded that the Spanner change stream makes it easier to guarantee at-lease-once delivery.
+
+### Open question
+
+For Phase 1:
+- How are we going to distinguish between `ObjectCreated:Put`, `ObjectCreated:Post`, `ObjectCreated:CompleteMultipartUpload`, and `ObjectCreated:Copy` events from Spanner change records.
+- What permissions do we need to set to Spanner and Pub/Sub?
+
+For future phases:
+- How do we provide these info to the notifications:
+  - `awsRegion`
+  - `userIndentity`
+  - `sourceIPAddress`
+  - `x-amz-request-id`
+  - `x-amz-id-2`
+  - `eTag`
+    - Can the `eTag` be decrypted with the satellite-managed passphrase?
```

Comment:
I think it should be possible based on some code-skimming, but I still don't have a complete understanding of how the eTag works